### PR TITLE
pythonPackages: use latest qt version instead of 5.14

### LINKED ISSUE #99937: Python packages depending on qt5, use qt514 while all-packages.nix defaults to qt515, causing incompatibilities.
**Describe the bug**

Experienced here: https://github.com/NixOS/nixpkgs/pull/99465#issue-497263839 and here: https://github.com/NixOS/nixpkgs/pull/99685#issuecomment-704782836

How come pyqt5 winds up with qt-5.14 in it's inputs and in it's references? Considering:

https://github.com/NixOS/nixpkgs/blob/433db454f3f6925addb1d7e26f134ef67567ac02/pkgs/top-level/all-packages.nix#L15002-L15003

And:

https://github.com/NixOS/nixpkgs/blob/433db454f3f6925addb1d7e26f134ef67567ac02/pkgs/top-level/python-packages.nix#L5278-L5280

**To Reproduce**

```
nix-store --query --references $(nix-build --no-out-link -A python3.pkgs.pyqt5) | grep 5.14
```

**Expected behavior**

No qt5.14 references.

**Screenshots**

**Additional context**

**Notify maintainers**

cc @sander

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: python.pkgs.pyqt5
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #99951: kdeconnect is broken in 20.09
**Describe the bug**

The kdeconnect widget in the KDE panel shows:

```
Error loading QML file: file:///nix/store/78r8ihcgk25cj91d5bk8l5z6xcvxqli9-kdeconnect-1.3.5/share/plasma/plasmoids/org.kde.kdeconnect/contents/ui/main.qml:25:1: module "org.kde.kdeconnect" is not installed
```

Also, going to the "KDE Connect" panel in the KDE System Settings shows the error message `The shared library was not found.`

**Notify maintainers**

@FRidh 

**Metadata**

```
$ nixos-version 
20.09.20201006.7badbf1 (Nightingale)
```

### LINKED ISSUE #99982: [20.09] electrum: qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
The Electrum wallet GUI fails with a coredump because the Qt platform is unable to load plugins.

**To Reproduce**
1. `nix run nixpkgs.electrum --command electrum`

**Expected behavior**
Electrum should launch without crashing.

**Screenshots**
```
nix run nixpkgs.electrum --command electrum
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted (core dumped)
```

**Additional context**
The electrum package already uses the Qt wrappers, and they worked fine at some point. I've compared the generated scripts to those of `qutebrowser`, which is working, and I didn't see any difference that could explain the problem.

**Notify maintainers**

@ehmry @joachifm @np

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.8.13, NixOS, 21.03.git.cfc9214a8ad (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - nixpkgs: `/home/emmanuel/projects/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: electrum
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

In https://github.com/NixOS/nixpkgs/commit/3cbcc14eef897607c50be4a9918d88614ea9881e the Qt version used in Python packages was pinned to 5.14. This was done because certain *applications* were not compatible. In principle, we should try to have the same Qt version in the Python packages set as in the top-level set, and override applications in case they're not compatible.

There were some problems with how the pinning to 5.14 was done, which eventually led to https://github.com/NixOS/nixpkgs/pull/98197.

This PR now unpins the Qt version. Applications are going to need some fixing, as is done for Qutebrowser in https://github.com/NixOS/nixpkgs/pull/98197.

This fixes https://github.com/NixOS/nixpkgs/issues/99937 + fixes https://github.com/NixOS/nixpkgs/issues/99982 + fixes #99951 .

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

# Packages

Just to note anyone wishing to help, these are the affected packages, filtered from ofborg's eval:

- [x] aj-snapshot
- [x] anki
- [x] ankisyncd
- [x] appleseed
- [x] asymptote
- [x] audiality2
- [x] backintime
- [x] backintime-qt4
- [x] cadence
- [x] calibre
- [x] calibre-py2 - build and run with `libsForQt5.callPackage` (instead of `libsForQt514.callPackage`) - not sure if this needs to be committed, considering https://github.com/NixOS/nixpkgs/commit/22167ae45fb27fac1e6aec7d1bf6ce82516fcb12 .
- [x] calibre-py3 - build and run with `libsForQt5.callPackage` (instead of `libsForQt514.callPackage`) - not sure if this needs to be committed, considering https://github.com/NixOS/nixpkgs/commit/22167ae45fb27fac1e6aec7d1bf6ce82516fcb12 .
- [x] carla
- [x] cq-editor
- [x] cura - The gui seems to work, though it segfaulted when I quit the program, and the console showed lots of deprecation warning. If qt5.14 will be desired for it, it will need a big diff for an override.
- [x] curaengine
- [x] curaLulzbot - same as cura
- [ ] dupeguru - fails to build from before this change ([example](https://hydra.nixos.org/build/127857616)).
- [x] electron-cash
- [x] electrum
- [x] electrum-ltc
- [x] faust2jack - only affected by sip patch, should be find if it builds.
- [x] faust2jaqt - only affected by sip patch, should be find if it builds.
- [x] fdroidserver
- [x] ffado - only affected by sip patch, should be find if it builds.
- [x] flatcam
- [x] flent
- [x] freecad
- [x] freecadStable
- [x] frescobaldi
- [x] friture - runs fine sort of, no qt related errors but does print opengl related errors.
- [x] giada
- [x] gitAndTools.git-annex-metadata-gui
- [x] gitAndTools.git-cola
- [x] git-cola
- [x] gmic-qt-krita
- [ ] gns3-gui - should be OK, as it's only pyqt5 that was upgraded for it.
- [x] gnss-sdr - should be tested with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gnuradio - should be tested with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gnuradio-with-packages - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gqrx - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-ais - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-gsm - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-limesdr - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-nacl - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-osmosdr - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] gr-rds - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [x] hplip
- [ ] hplip_3_16_11
- [ ] hplip_3_18_5
- [x] hplipWithPlugin
- [ ] hplipWithPlugin_3_16_11
- [ ] hplipWithPlugin_3_18_5
- [ ] ihaskell
- [x] inkcut
- [x] inspectrum - should be unaffected with https://github.com/NixOS/nixpkgs/pull/99685
- [ ] jack1
- [ ] jack2
- [ ] jack2Full
- [ ] jackmix_jack1
- [x] kcc
- [x] kmymoney
- [ ] koboredux
- [ ] koboredux-free
- [x] krita
- [x] krop
- [ ] labelImg - - should be OK, as it's only pyqt5 that was upgraded for it.
- [x] leo-editor
- [x] libffado
- [x] littlegptracker
- [x] lsp-plugins
- [x] ltc-tools
- [x] luppp
- [x] maestral-gui (https://github.com/NixOS/nixpkgs/pull/99589)
- [x] manuskript
- [ ] mnemosyne
- [ ] mooSpace
- [ ] multibootusb
- [ ] nagstamon
- [x] onionshare
- [x] onionshare-gui
- [x] openshot-qt
- [x] persepolis
- [X] picard - should be OK
- [ ] plover.dev
- [x] puddletag (`qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
`)
- [x] pulseeffects
- [x] python27Packages.androguard
- [ ] python27Packages.appleseed
- [ ] python27Packages.binwalk-full
- [ ] python27Packages.jupyter
- [ ] python27Packages.pivy
- [ ] python27Packages.poppler-qt5
- [ ] python27Packages.pykdl
- [ ] python27Packages.pyqt4
- [x] python27Packages.pyqt5
- [x] python27Packages.pyqt5_with_qtmultimedia
- [ ] python27Packages.pyqt5_with_qtwebkit
- [ ] python27Packages.pyqtgraph
- [ ] python27Packages.pyqtwebengine
- [ ] python27Packages.pyside2
- [ ] python27Packages.pyside2-tools
- [ ] python27Packages.qscintilla
- [x] python27Packages.qscintilla-qt4
- [ ] python27Packages.qscintilla-qt5
- [ ] python27Packages.qtconsole
- [ ] python27Packages.shiboken2
- [x] python27Packages.sip
- [ ] python27Packages.subdownloader
- [x] python3Packages.androguard
- [ ] python3Packages.ansible-kernel
- [ ] python3Packages.binwalk-full
- [ ] python3Packages.enaml
- [ ] python3Packages.enamlx
- [ ] python3Packages.jupyter
- [ ] python3Packages.libarcus
- [ ] python3Packages.libsavitar
- [ ] python3Packages.lightparam
- [ ] python3Packages.mayavi
- [ ] python3Packages.pivy
- [ ] python3Packages.poppler-qt5
- [ ] python3Packages.pykdl
- [x] python3Packages.pyqt4
- [x] python3Packages.pyqt5
- [x] python3Packages.pyqt5_with_qtmultimedia
- [x] python3Packages.pyqt5_with_qtwebkit
- [ ] python3Packages.pyqtgraph
- [ ] python3Packages.pyqtwebengine
- [ ] python3Packages.pyside2
- [ ] python3Packages.pyside2-tools
- [ ] python3Packages.qimage2ndarray
- [ ] python3Packages.qreactor
- [ ] python3Packages.qscintilla-qt5
- [ ] python3Packages.qtconsole
- [ ] python3Packages.quamash
- [ ] python3Packages.roboschool
- [ ] python3Packages.shiboken2
- [x] python3Packages.sip
- [ ] python3Packages.skein
- [x] python3Packages.spyder
- [ ] python3Packages.spyder_3 (`qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.`, old version, probably best removed)
- [ ] python3Packages.stytra
- [ ] python3Packages.trezor_agent
- [ ] python3Packages.uranium
- [ ] python3Packages.vega
- [x] python3Packages.weboob - builds, but I couldn't find a single binary that should launch a gui there.
- [x] qarte
- [x] qgis (`error: no matching function for call to ‘QsciAPIs::QsciAPIs()’`)
- [x] qgis-unwrapped
- [x] qnotero
- [x] qpaeq
- [x] qradiolink - should be unaffected / tested along with https://github.com/NixOS/nixpkgs/pull/99685
- [x] qutebrowser - Should work, given that qt is not updated for it ([commit](https://github.com/NixOS/nixpkgs/pull/99956/commits/2667af4062ef0710b486fbae3108e0141aecb0fc)).
- [x] rapid-photo-downloader
- [x] retext
- [ ] scudcloud
- [ ] seexpr
- [ ] shotcut
- [ ] sl1-to-photon
- [ ] sonic-pi
- [x] soundtracker
- [x] spyder
- [x] sumorobot-manager
- [x] syncplay
- [ ] tambura
- [ ] tetraproc
- [x] tortoisehg
- [ ] trezor_agent
- [ ] tribler
- [x] tuxguitar
- [ ] urh
- [ ] virtscreen
- [x] vorta
- [x] webbrowser
- [ ] webcamoid
- [x] webmacs

### CODE REVIEW

**Comment by FRidh** [comment]
It's important here that we build and test at least some Python+Qt applications.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10213,10 +10213,9 @@ in
   python3Packages = python3.pkgs;
 
   pythonInterpreters = callPackage ./../development/interpreters/python {
-    # Overrides that apply to all Python interpreters
+    # Overrides that apply to all Python interpreters and their packages
+    # Generally, this should be avoided.
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:10217]: :green_heart:

**Comment by doronbehar** [comment]
I have a question regarding qutebrowser. Why isn't like this:

```diff
diff --git i/pkgs/top-level/all-packages.nix w/pkgs/top-level/all-packages.nix
index c13b60c5d80..73300080f66 100644
--- i/pkgs/top-level/all-packages.nix
+++ w/pkgs/top-level/all-packages.nix
@@ -22990,7 +22990,8 @@ in
     pkgs_ = pkgs.extend(_: prev: {
       pythonInterpreters = prev.pythonInterpreters.override(oldAttrs: {
         pkgs = oldAttrs.pkgs.extend(_: _: {
-          inherit (pkgs) qt5 libsForQt514;
+          qt5 = qt514;
+          libsForQt5 = libsForQt514;
         });
       });
     });
```

**Comment by FRidh** [comment]
> I have a question regarding qutebrowser. Why isn't like this:

There are two parts:
1. me (?) being sloppy, using a versioned *and* unversioned attribute.
2. `pkgs` was the one originally passed into the Python packages set. Since we're overriding, I am of the opinion we should aim at picking items from that set. In practice I suppose it does not matter much.

**Comment by doronbehar** [comment]
According to qutebrowser [latest release notes](https://github.com/qutebrowser/qutebrowser/releases/tag/v1.13.1), qt5.15 is supported. I'm confused now - https://github.com/NixOS/nixpkgs/pull/97586 tried to make it use qt5.15, why won't it use 5.15 now?

**Comment by FRidh** [comment]
> According to qutebrowser latest release notes, qt5.15 is supported. I'm confused now - #97586 tried to make it use qt5.15, why won't it use 5.15 now?

Indeed, see https://github.com/NixOS/nixpkgs/pull/97586. However.....see also 3fafb021256b. So the sloppiness wasn't mine here ;)

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -22990,7 +22990,10 @@ in
     pkgs_ = pkgs.extend(_: prev: {
       pythonInterpreters = prev.pythonInterpreters.override(oldAttrs: {
         pkgs = oldAttrs.pkgs.extend(_: _: {
-          inherit (pkgs) qt5 libsForQt514;
+          # Use 5.14
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:22993]:
```suggestion
          # Use 5.14 (https://github.com/qutebrowser/qutebrowser/issues/5359)
```

**Comment by doronbehar** [comment]
There are many `libsForQt514.callPackage` attributes in all-packages.nix . Some of which call expressions that import `pythonPackages` which then use `pyqt5` and that pyqt5 is built with qt5.15.. Most of these pinnings were made in https://github.com/NixOS/nixpkgs/commit/22167ae45fb27fac1e6aec7d1bf6ce82516fcb12 , not sure where @ttuegel got the list of what broke with qt5.15.

In any case, If it'll turn out that we can't switch to `libsForQt5.callPackage` for these packages, I'm afraid we'll wind up with many more ugly `qutebrowser` like overrides.. Do you think @FRidh we could make that a bit prettier?

**Comment by FRidh** [comment]
> In any case, If it'll turn out that we can't switch to libsForQt5.callPackage for these packages, I'm afraid we'll wind up with many more ugly qutebrowser like overrides.. Do you think @FRidh we could make that a bit prettier?

I suppose one could do
```nix
mypackage = (pkgs.extend(final: prev: {
  qt5 = prev.qt514;
  libsForQt5 = prev.libsForQt514;
})).libsForQt5.callPackage ... { };

```
which I think is cleanest. This probably takes more effort to evaluate though.

File: pkgs/development/python-modules/pyqtwebengine/default.nix [hunk 3]
```diff
@@ -7,6 +7,12 @@ let
 
   inherit (pythonPackages) buildPythonPackage python isPy3k pyqt5 enum34;
   inherit (pyqt5) sip;
+  patches = (if lib.hasPrefix "5.14" pyqt5.version then
```
Comment (FRidh) [inline pkgs/development/python-modules/pyqtwebengine/default.nix:10]: `lib.optional`

**Comment by SFrijters** [comment]
maestral-gui is fixed by this PR.

**Comment by doronbehar** [comment]
> maestral-gui is fixed by this PR.

@SFrijters it's a bit surprising to hear. It seems to me that we should have included the changes in https://github.com/NixOS/nixpkgs/pull/99589 here, or at least it should be safer to do so.

**Comment by doronbehar** [comment]
> > In any case, If it'll turn out that we can't switch to libsForQt5.callPackage for these packages, I'm afraid we'll wind up with many more ugly qutebrowser like overrides.. Do you think @FRidh we could make that a bit prettier?
> 
> I suppose one could do
> 
> ```nix
> mypackage = (pkgs.extend(final: prev: {
>   qt5 = prev.qt514;
>   libsForQt5 = prev.libsForQt514;
> })).libsForQt5.callPackage ... { };
> ```
> 
> which I think is cleanest. This probably takes more effort to evaluate though and thus probably should be avoided.

What if we'll put a `pyqt5-py3` attribute in every `libsForQt5` attrset? This way, in `all-packages.nix` you could simply `libsForQt514.callPackage` and then in the expression import `pyqt5-py3` directly.

`cadence` for example needs a similar override to `qutebrowser`.

**Comment by FRidh** [comment]
> What if we'll put a pyqt5-py3 attribute in every libsForQt5 attrset? This way, in all-packages.nix you could simply libsForQt514.callPackage and then in the expression import pyqt5-py3 directly.

No, definitely not that. That's regressing to the situation we had years ago, when we had Python libraries of different versions in the top-level. In that case I would be more in favor of having `python3` in `libsForQt5` with that version of Qt, but that's also not really a solution, because it would mean the entire dependency tree would use that `python` with Qt.

Overriding, while a feature, is an ugly system. Often it seems fine, but here we see the limitation of the current implementation.

But, so far, it does not seem like there are many packages that need the override, right?

**Comment by gebner** [comment]
> > According to qutebrowser latest release notes, qt5.15 is supported. I'm confused now - #97586 tried to make it use qt5.15, why won't it use 5.15 now?
> 
> Indeed, see #97586. However.....see also [3fafb02](https://github.com/NixOS/nixpkgs/commit/3fafb021256bc594cecd949b3edc5bc480fc721f). So the sloppiness wasn't mine here ;)

Note that the upstream author strongly recommends using Qt 5.15 for qutebrowser (because of the obvious security issues when using old browser engines): https://github.com/NixOS/nixpkgs/pull/99456#issuecomment-703523943

**Comment by FRidh** [comment]
testing an upgrade of qscintilla. I think it is needed for qgis.

**Comment by jorsn** [comment]
Here is a nix function for checking the version consistency of package dependencies: https://paste.sr.ht/~jorsn/545730b6ab00e13110b1dd94dd7031be916425cd.

**Comment by FRidh** [comment]
Build status as of several commits ago. Of course the real question is run-time.
```
41 packages failed to build:
appleseed dupeguru freecadStable gnss-sdr gnuradio gr-ais gr-gsm gr-limesdr gr-nacl gr-osmosdr gr-rds gnuradio-with-packages gqrx ihaskell inspectrum koboredux nagstamon openshot-qt python27Packages.ovito python27Packages.pykdl python27Packages.pyqt5_with_qtwebkit python27Packages.qscintilla python27Packages.qscintilla-qt5 python37Packages.ovito python37Packages.pykdl python37Packages.pyqt5_with_qtwebkit python37Packages.qscintilla-qt5 python37Packages.skein python38Packages.ovito python38Packages.pykdl python38Packages.pyqt5_with_qtwebkit python38Packages.qscintilla-qt5 python38Packages.skein qgis qgis-unwrapped qradiolink qutebrowser scudcloud seexpr tortoisehg webbrowser                                            

170 packages built:
aj-snapshot anki ankisyncd asymptote audiality2 backintime cadence calibre calibre-py2 carla cde-gtk-theme cq-editor cura curaLulzbot curaengine electron-cash electrum electrum-ltc faust2jack faust2jaqt fdroidserver ffado flatcam flent freecad frescobaldi friture giada git-cola gitAndTools.git-annex-metadata-gui gmic-qt-krita gns3-gui hplip hplipWithPlugin hplipWithPlugin_3_16_11 hplipWithPlugin_3_18_5 hplip_3_16_11 hplip_3_18_5 inkcut jack1 jack2 jackmix_jack1 kcc kmymoney koboredux-free krita krop labelImg leo-editor littlegptracker lsp-plugins ltc-tools luppp maestral-gui manuskript mnemosyne mooSpace multibootusb onionshare onionshare-gui persepolis picard plover.dev puddletag pulseeffects python27Packages.androguard python27Packages.binwalk-full python27Packages.jupyter python27Packages.nose-timer python27Packages.pivy python27Packages.poppler-qt5 python27Packages.pyqt4 python27Packages.pyqt5 python27Packages.pyqt5_with_qtmultimedia python27Packages.pyqtgraph python27Packages.pyqtwebengine python27Packages.pyside2 python27Packages.pyside2-tools python27Packages.qtconsole python27Packages.shiboken2 python27Packages.sip python27Packages.subdownloader python37Packages.androguard python37Packages.ansible-kernel python37Packages.binwalk-full python37Packages.enaml python37Packages.enamlx python37Packages.jupyter python37Packages.libarcus python37Packages.libsavitar python37Packages.lightparam python37Packages.mayavi python37Packages.nose-timer python37Packages.pivy python37Packages.poppler-qt5 python37Packages.pyqt4 python37Packages.pyqt5 python37Packages.pyqt5_with_qtmultimedia python37Packages.pyqtgraph python37Packages.pyqtwebengine python37Packages.pyside2 python37Packages.pyside2-tools python37Packages.qimage2ndarray python37Packages.qreactor python37Packages.qtconsole python37Packages.quamash python37Packages.roboschool python37Packages.shiboken2 python37Packages.sip python37Packages.spyder python37Packages.spyder_3 python37Packages.stytra python37Packages.trezor_agent python37Packages.uranium python37Packages.vega python37Packages.weboob python38Packages.androguard python38Packages.ansible-kernel python38Packages.binwalk-full python38Packages.enaml python38Packages.enamlx python38Packages.jupyter python38Packages.libarcus python38Packages.libsavitar python38Packages.lightparam python38Packages.mayavi python38Packages.nose-timer python38Packages.pivy python38Packages.poppler-qt5 python38Packages.pyqt4 python38Packages.pyqt5 python38Packages.pyqt5_with_qtmultimedia python38Packages.pyqtgraph python38Packages.pyqtwebengine python38Packages.pyside2 python38Packages.pyside2-tools python38Packages.qimage2ndarray python38Packages.qreactor python38Packages.qtconsole python38Packages.quamash python38Packages.roboschool python38Packages.shiboken2 python38Packages.sip spyder python38Packages.spyder_3 python38Packages.stytra trezor_agent python38Packages.uranium python38Packages.vega python38Packages.weboob qarte qnotero qpaeq rapid-photo-downloader retext shotcut sl1-to-photon sonic-pi soundtracker sumorobot-manager syncplay tambura tetraproc tribler tuxguitar urh virtscreen vorta webcamoid webmacs
```

**Comment by doronbehar** [comment]
Technically we don't have to build them all. Somehow jack packages got changed outputs and that triggerred many rebuilds I can't explain.

I'm dictating now the policy where only if I change `libsForQt{514 -> 5}` I test the rebuild, and if it's only pyqt5 that has changed (because now it builds with qt5.15 of course), I don't test it. Some packages get bonus fixes such as double wrapping prevented.

As for that report: 

- Many gnuradio related should not worry as, for we have https://github.com/NixOS/nixpkgs/pull/99685, as explained in the big list at the top.
- webbrowser builds on hydra, so I'm surprised it failed, testing now.
- Some packages won't even evaluate so it shouldn't block the PR.
- Some packages fail to build on hydra as well, with unrelated errors - not blocking as well.

**Comment by jorsn** [comment]
> @jorsn do you think that can be incorporated into `libsForQt5.callPackage`?

It takes too much time, and I think you cannot probably optimize it sucht that this gets neglectible:

~~~
$ time nix eval --impure --expr 'with import ./. {}; import ./dep-ver-consistent.nix lib false 1 "qtbase" carla'      
trace: x.pkg.version == 5.15.0 != 5.14.2 == dep0.pkg.version
  where
    dep0.parent.outPath == /nix/store/xmgiwrnglzmkyyz2ccg54jjsswjhayzd-carla-2.1.1
    dep0.pkg.outPath == /nix/store/vm9m827nfcnlf97y53bn65zaadw8pasp-qtbase-5.14.2-dev
    x.parent.outPath == /nix/store/7414sivcqpizf81pjrj99qbib9wx50kk-python3.8-PyQt5-5.15.1-dev
    x.pkg.outPath == /nix/store/njywy8s62w93ymr822sc9cl03nv72q7r-qtbase-5.15.0-dev

false
nix eval --impure --expr   1.21s user 0.19s system 84% cpu 1.650 total
~~~

But it could be an optional check.

My function is not really optimized: It checks packages whenever they occur, so it also checks duplicates doubly. This shoudln't be a big problem if one checks shallowly, and depth 1 is sufficient if pyqt5 using another qt version is a direct dependency.

**Comment by doronbehar** [comment]
Hmm.. Maybe it should be an ofborg only check. If others will agree it's a good idea.

**Comment by FRidh** [comment]
Our qgis version (3.10, LTS) is compatible with 5.15, however, it also *requires* qtwebkit, and qtwebkit on 5.15 fails to build according to 01bcfd3579219d60e5d07df309a000f96b2b658b.
Using Qt 5.14 should solve the issue. It requires an override like for qutebrowser.

**Comment by doronbehar** [comment]
webbrowser is fine.

**Comment by SFrijters** [comment]
@doronbehar 

> > maestral-gui is fixed by this PR.
> 
> @SFrijters it's a bit surprising to hear. It seems to me that we should have included the changes in #99589 here, or at least it should be safer to do so.

Why do you think this is surprising? As far as I understand it, the problem was caused by the top-level `libsForQt` using 5.15 and `pyqt5` using 5.14, so my fix was to pin it back to `libsForQt514`. But if `pyqt5` now uses 5.15, it's all consistent, right?
Still, it does feel a bit fragile, if the top-level Qt version and the pythonPackages Qt version go out of sync?
I've skimmed through this discussion, but I don't know enough about the internals of the package sets to make a useful contribution.

**Comment by FRidh** [comment]
> Still, it does feel a bit fragile, if the top-level Qt version and the pythonPackages Qt version go out of sync?

Normally they won't get out of sync, it only happened here as part of reducing the scope of the Qt/KDE upgrades.

**Comment by doronbehar** [comment]
It's OK, from some reason I thought it was `libsForQt514` not in this `qt` branch.

**Comment by doronbehar** [comment]
I've finally finished iterating all packages, not necessarily building all, but the remaining work I see to be done is:

1. Figure out why jack packages were affected - a big number of the unmarked packages in the list above is packages affected only by jack.
2. Consider removing the override for qutebrowser.
3. Fix an override for:
   - cadence
   - qgis
4. Squash `fixup!` commits.

I could use some help with (1). I'm not sure about (2), and for (3) I'm not sure whether to use [this](https://github.com/NixOS/nixpkgs/pull/99956#issuecomment-705019945) version of the override, or the known one currently used for qutebrowser.

**Comment by FRidh** [comment]
1. squashed the fixup commit
2. qgis now works

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -19741,7 +19741,17 @@ in
 
   bambootracker = libsForQt5.callPackage ../applications/audio/bambootracker { };
 
-  cadence = libsForQt514.callPackage ../applications/audio/cadence { };
+  cadence = let
+    # Use Qt 5.15 consistently
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:19745]: 5.14

**Comment by doronbehar** [comment]
jack packages were affected by `ffado` which now use a consistent qt version so I consider them all to be OK.

**Comment by FRidh** [comment]
What is the problem with the jack packages? Tried several, and they build, and jackmix ran as well.

**Comment by doronbehar** [comment]
No problem at all, just wondering why their inputs got changed. ffado needed a consistent qt version commit though.

**Comment by doronbehar** [comment]
:1+: on dupeguru.

**Comment by doronbehar** [comment]
I think we should remove the qutebrowser override. We don't like it, and upstream recommends not doing it.

**Comment by doronbehar** [comment]
I also think we should merge the qt5.15 patch upstream mentioned: https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/315691

**Comment by FRidh** [comment]
> I think we should remove the qutebrowser override. We don't like it, and upstream recommends not doing it.

I agree, see 902c567974a3af0795c2efce369dc0be0edea34b and https://github.com/NixOS/nixpkgs/pull/99456#issuecomment-705708931.

**Comment by FRidh** [comment]
The `python*Packages` typically won't work out of the box, because they require wrapping (https://github.com/NixOS/nixpkgs/pull/83321).

**Comment by doronbehar** [comment]
@FRidh maybe this is not related, but I'm a bit worried about kdeconnect's ld errors I got in https://github.com/NixOS/nixpkgs/issues/99951#issuecomment-705635228 it seems as if mixed version of qt were used, though I couldn't find any. Gonna try @jorsn's script.

**Comment by doronbehar** [comment]
That was helpful:

```
>> $(nix-build -A nixUnstable)/bin/nix eval --impure --expr 'with import ./. {}; import ./dep-ver-consistent.nix lib false 1 "qtbase" kdeconnect'
trace: x.pkg.version == 5.14.2 != 5.15.0 == dep0.pkg.version
  where
    dep0.parent.outPath == /nix/store/acgqlww61g8nmcyqn96fvdir8fgd9h6a-qtmultimedia-5.15.0-dev
    dep0.pkg.outPath == /nix/store/njywy8s62w93ymr822sc9cl03nv72q7r-qtbase-5.15.0-dev
    x.parent.outPath == /nix/store/7bxbjcp3j3rgj4g81vq91kqk7vz00i8y-kdoctools-5.73.0-dev
    x.pkg.outPath == /nix/store/vm9m827nfcnlf97y53bn65zaadw8pasp-qtbase-5.14.2-dev
```

Though it's not easy to trace it...

**Comment by doronbehar** [comment]
Found the culprit: Both kapidox and kdoctools _do_ in fact depend and reference qt:

https://github.com/NixOS/nixpkgs/blob/0790b3e4dc797a097175eb14361dab32217c1a00/pkgs/top-level/all-packages.nix#L10824-L10825

And `kdeconnect` used them instead of their versions inside it's `libsForQt5` attribute set.

**Comment by doronbehar** [comment]
That still won't help kdeconnect, but at least now the missing header file is consistent for when using qt5.14 and qt5.15.

**Comment by FRidh** [comment]
Anything more? From my point of view this is ready.

**Comment by doronbehar** [comment]
Agree :vulcan_salute: .

**Comment by FRidh** [comment]
Now we need to look into cherry-picking fixes to 20.09, depending on what Qt versions were used there.

**Comment by doronbehar** [comment]
I think all of the changes need to be cherry-picked because all of the commits removed the use of non defaults - i.e `libsForQt{514 -> 5}`. Hence whatever is the default on all-packages.nix and python-packages.nix on 20.09, should both fit each other.

**Comment by piegamesde** [comment]
Hm, this does not fix it for me. `NIX_PATH=nixpkgs=https://github.com/nixos/nixpkgs/archive/master.tar.gz nix-shell -p 'electrum' --run electrum` still has the same error. Kile fails as well, on the other side Git-cola works.

**Comment by doronbehar** [comment]
Kile was not affected by this issue, since it does not use pyqt5. So it is unrelated.

**Comment by jorsn** [comment]
Kile is no python package, so outside the scope of this issue.

In Principle, the problem is the same as the former python problem: `kdeApplications` and `kdeFrameworks` use qt514 (see their definitions in `pkgs/top-level/all-packages.nix`) and thus any package using `qt5` == `qt515` and `kdeFrameworks` have inconsistent qt5 versions in their scope.

Kile can simply be fixed by using `libsForQt514.callPackage` instead of `libsForQt5.callPackage`, or by updating kde to use qt 5.15 as well, analogously to this issue.

**Comment by jorsn** [comment]
I agree. `electrum` is also wrapped doubly.

**Comment by FRidh** [comment]
Working now on cherry-picking this to 20.09.

**Comment by zimbatm** [comment]
Please stop using `pkgs.extend`, it blows up the memory usage of 130MB on each instance!

**Comment by doronbehar** [comment]
> Please stop using pkgs.extend, it blows up the memory usage of 130MB on each instance!

Interesting metric. We should be able to get rid of it in https://github.com/NixOS/nixpkgs/pull/104474 .

**Pull request merged**